### PR TITLE
Fix/master/testing with the help of docker

### PR DESCRIPTION
**Pull request by anestos**
Changed the way testing was done to not rely on a Curity server to be running nor the httpbin.org website. 
To run the tests do the following:

- In the `docker-compose.yaml` file add a valid `LICENCE_KEY` and an `ADMIN_PASSWORD` in the environment variables
- `./configure` (select **Yes** for Debug features and  **No** for Dynamic module)
- `make`
- `make test`

docker is required and must be running in your machine

### CODE REVIEW

File: docker-compose.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+services:
+  curity-service:
+    image: curity/idsvr
+    ports:
+      - 6749:6749
+      - 8443:8443
+    environment:
+      - ADMIN=true
+      - SERVICE_ROLE=default
+      - LICENSE_KEY=${LICENSE_KEY:?err}
```
Comment (travisspencer) [inline docker-compose.yaml:11]: these have to be set manually? From the README above:

> edit the `docker-compose.yaml` file and configure the variables `ADMIN_PASSWORD` and `LICENSE_KEY`

They can't be set in the environment when `docker-compose` is run?

**Review by travisspencer — APPROVED** [review]

**Comment by travisspencer** [comment]
When I run the tests, I get these failures:
```
not ok 7 - TEST 5: Module can be disabled at the HTTP level - pattern "Module disabled" should match a line in error.log (req 0)

#   Failed test 'TEST 5: Module can be disabled at the HTTP level - pattern "Module disabled" should match a line in error.log (req 0)'
#   at /Users/tspencer/perl5/lib/perl5/Test/Nginx/Socket.pm line 1213.
# Looks like you failed 1 test of 8.
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/8 subtests

t/curity.t ..
not ok 1 - Could not get token from idsvr

#   Failed test 'Could not get token from idsvr'
#   at t/curity.t line 16.
# Looks like you failed 1 test of 1.
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests

Test Summary Report
-------------------
t/config.t (Wstat: 256 Tests: 8 Failed: 1)
  Failed test:  7
  Non-zero exit status: 1
t/curity.t (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
```

File: README.md [hunk 2]
```diff
@@ -287,15 +287,18 @@ Pre-built binaries of this module are provided for the following versions of NGI
...
-* An introspection endpoint with a URI of `/introspection`;
-* An OAuth client named `client-one` with a secret of `0ne!Secret` and the client credential capability; and
-* An OAuth client named `test_gateway_client` and a secret of `Password1` with the introspection capability.
+To run the tests do the following:
 
-Internet access to `httpbin.org` is required for the `curity.t` test suite to pass.
+* run `./configure`, make sure you select _Yes_ for Debug features and _No_ for Dynamic module 
+* run `make`
+* edit the `docker-compose.yaml` file and configure the variables `ADMIN_PASSWORD` and `LICENSE_KEY`
+* run `make test`
```
Comment (travisspencer) [inline README.md:297]: we should probably say how to stop the containers after the tests

File: docker-compose.yaml (hunk 1 above)
Comment (anestos) [reply docker-compose.yaml:11]: Yeah they can be environment variables

**Comment by anestos** [comment]
> When I run the tests, I get these failures:
> 
> ```
> not ok 7 - TEST 5: Module can be disabled at the HTTP level - pattern "Module disabled" should match a line in error.log (req 0)
> 
> #   Failed test 'TEST 5: Module can be disabled at the HTTP level - pattern "Module disabled" should match a line in error.log (req 0)'
> #   at /Users/tspencer/perl5/lib/perl5/Test/Nginx/Socket.pm line 1213.
> # Looks like you failed 1 test of 8.
> Dubious, test returned 1 (wstat 256, 0x100)
> Failed 1/8 subtests
> 
> t/curity.t ..
> not ok 1 - Could not get token from idsvr
> 
> #   Failed test 'Could not get token from idsvr'
> #   at t/curity.t line 16.
> # Looks like you failed 1 test of 1.
> Dubious, test returned 1 (wstat 256, 0x100)
> Failed 1/1 subtests
> 
> Test Summary Report
> -------------------
> t/config.t (Wstat: 256 Tests: 8 Failed: 1)
>   Failed test:  7
>   Non-zero exit status: 1
> t/curity.t (Wstat: 256 Tests: 1 Failed: 1)
>   Failed test:  1
>   Non-zero exit status: 1
> ```

I am getting the same only if during `./configure` I don't enable debug features. Is it the same for you or you had debug on?

File: docker-compose.yaml (hunk 1 above)
Comment (travisspencer) [reply docker-compose.yaml:11]: Can you add an example to the docs on how to pass them in that way? I tried `ADMIN_PASSWORD=foo LICENCE_KEY=bar make test` and `make test ADMIN_PASSWORD=foo LICENCE_KEY=bar` and `export ADMIN_PASSWORD=foo LICENCE_KEY=bar ; make test`. None of those seemed to work for me.

**Comment by travisspencer** [comment]
> I am getting the same only if during `./configure` I don't enable debug features.

In that case `TEST 5` passes, but the Curity one still doesn't work for me.

File: docker-compose.yaml (hunk 1 above)
Comment (anestos) [reply docker-compose.yaml:11]: All of those ways work for me, you can also add the values in a .env file. I ll add a note in the comments

File: README.md (hunk 2 above)
Comment (anestos) [reply README.md:297]: They are stopped after the tests run

**Comment by anestos** [comment]
> > I am getting the same only if during `./configure` I don't enable debug features.
> 
> If I configure with debug symbols, the case `TEST 5` passes, but the Curity one still doesn't work for me.

Any chance you had Curity runnning locally so the request for the token went there instead of in the docker container?

**Comment by travisspencer** [comment]
> Any chance you had Curity runnning locally

No. I got an error from `docker-compose` at first, and then shut it down. So, I'm sure it's not running locally.

File: Makefile [hunk 3]
```diff
@@ -20,7 +20,11 @@ clean:
 	rm -rf .build.info nginx-$(NGINX_VERSION) nginx-$(NGINX_VERSION).tar.gz* t/servroot
 
 test: all
-	PATH=$(NGINX_SRC_DIR)/objs:$$PATH prove -v t/		
+	docker-compose up -d
+	@echo "Waiting for the Curity Identity Server to start..."
+	@bash -c 'c=0; while [[ $$c -lt 25 && "$$(curl -fs -w ''%{http_code}'' localhost:8443)" != "404" ]]; do ((c++)); echo -n "."; sleep 1; done'
```
Comment (travisspencer) [inline Makefile:25]: @anestos , I added this.

File: t/config.t [hunk 4]
```diff
@@ -84,6 +85,8 @@ GET /t
 --- error_log
 Module disabled
 
+--- skip_eval: 1: open(FH, "<", "$FindBin::Bin/../.build.info"); my $skip=0; while (<FH>) { $skip = 1 if ($_ =~ /DEBUG=n/) } $skip
```
Comment (travisspencer) [inline t/config.t:88]: added this @anestos

**Pull request merged**